update-type, remove-type, change recognition

tobkle opened the pull request:
Enhanced create-graphql-server with an "update-type", "remove-type" procedure. 

If a file was changed by a user, which will be overwritten by any "*-type" command, it warns and stops. If the user decides, that it is ok to overwrite, he can restart the previous command with a "force-update" option to overrule the check. 

The file change recognition works with a simple md5 hash calculation, which is stored in a file 'GraphQL.checksums'. In the beginning it is not there, it will be created by the first "add-type" run. 

Added some colors to the CLI user interface to recognize errors more easily. 

Added a test with a new "test/test-add-update-remove.sh" script. 

This test creates two temporary directories and does a CGS init into both of them. In the first directory it does several changes with 3x add-type, then an update-type, an update-type with file change, an update-type with file-change and "force-update" option, and finally it does 3x remove-type. Then the first and the second directory are compared with each other, as they have to be equally again.

The 5 tests check for generated file existences and if the index files contains the type string which has the type file loaded.

Please have a look, if it fits to your project principles and quality expectations.

tobkle commented:
Somehow I can't see the ./bin/create-graphql-server.sh file listed in the PR files. Maybe it doesn't come over with the merge. In the commits it was included. Do you have an idea why it is missing?

tmeasday commented on CHANGELOG.md:
```diff
@@ -1,3 +1,23 @@
+ - Added CMD option: remove-type to move existing versions into there respective archive folders
+ - Copied the two examples User.graphql, Tweet.graphql into the input folder as examples
+ - Added one npm dependency 'escape-string-regexp' for escaping code lines to be used in regular expressions
+
```
Can coalesce these into a simple changelog that just describes the user facing changes. (Plus you can use git to squash your changes if you want to clean things up, lmk if I can help with that).

tobkle replied:
Hm, then I assume you pipe a git log somehow while committing changes. But which options are you using therefore? The list I'm getting is too big.

tmeasday replied:
Let's just manually write the changelog. It's user visible, so it's best to just make something that is easy for users to understand that summarizes the changes.

tmeasday replied:
(I mentioned git squash as a separate thing; if you want to clean up your commits--I'm not stressed if you don't--you can use squash for it.)

tmeasday replied:
(i.e. you could remove the commits that add the archive logic if you want)

tobkle replied:
done

tmeasday commented on README.md:
````diff
@@ -5,7 +5,7 @@ This is a simple scaffolding tool for GraphQL apps, built on MongoDB.
 
 ```bash
-create-graphql-server init project-dir
+create-graphql-server init project-dir option
````
What is `option`?

tobkle replied:
removed it, wrong place

tmeasday commented on README.md:
````diff
@@ -70,6 +70,30 @@ If the field references an array (again w/ or w/o nullability) of another type,
+To update types use the following command:
+
+```bash
+create-graphql-server update-type path/to/input.graphql
````
Maybe `create-graphql-server update-type [--force-update] path/to/input.graphql`?

tobkle replied:
done

tmeasday commented on test/test-add-update-remove.sh:
```diff
@@ -0,0 +1,132 @@
+    grep -q "User" "./model/index.js"  &&
+    grep -q "Tweet" "./model/index.js" && 
+    grep -q "Order" "./model/index.js" &&  
+    grep -q "User" "./resolvers/index.js" && 
```
Should we check these appear the right number of times?

tobkle replied:
Good idea, did it this way:
[ `grep -o "User" "./model/index.js" | wc -l` -eq "1" ]

tobkle replied:
changed it into functions, and inform tester about which file is missing/wrong

tmeasday commented on bin/create-graphql-server.js:
```diff
@@ -8,29 +8,348 @@ import cpr from 'cpr';
+  console.log('');
+  console.log(chalk.yellow(' Options:'));
+  console.log('');
+  console.log(' force-update   forces updates, overwriting code changes by any user since last run')
```
Let's make this an option, i.e. `--force-update`

tobkle replied:
done

tmeasday commented on bin/create-graphql-server.js:
```diff
@@ -8,29 +8,348 @@ import cpr from 'cpr';
 const commands = argv._;
-
 const SKEL_DIR = path.join(__dirname, '..', 'skel');
+const CHECKSUM_FILE = path.join(__dirname, 'GraphQL.checksums');
```
I think maybe `.create-graphql-server.checksums`?

tmeasday replied:
(Note leading `.` to make it a hidden file)

tmeasday replied:
Also this is relative to the script (`__dirname`). I think you want relative to the cwd.

tobkle replied:
both done

tmeasday commented:
Also, I wonder if we should check the server continues to run after each step in the process?

tmeasday commented:
Also can we add the test to the circle.ci

tmeasday commented:
I just tested it out and it works great! Such a bit improvement.

Can you ensure this passes linting? I think I have been generally using [prettier](https://github.com/prettier/prettier) for formatting recently. I should setup `eslint-prettier` on this project (although it's annoying because the `skel` does not pass linting).

tmeasday commented on bin/create-graphql-server.js:
```diff
@@ -8,29 +8,348 @@ import cpr from 'cpr';
+console.log(chalk.bold.blue('CREATE-GRAPHQL-SERVER'));
+console.log('');
+
+let force_update = false;
```
Can we pass this in as a variable rather than using a global?

tobkle replied:
done

tmeasday commented:
Psyched about this. cc @shilman

tobkle commented:
Changed all, according to your suggestions, except for:
* Automatic Changelog.md generation => Can you give some hints, how you are using git log?
* Test Script Server restarts => Agree, but I wonder how to start/test/stop the server by the script

```javascript
process.title = 'GraphQL-Server';

"scripts": {
    "start": "node server/index.js",
    "stop": "pkill --signal SIGINT 'GraphQL-Server'"
}
```

tmeasday commented:
> Test Script Server restarts => Agree, but I wonder how to start/test/stop the server by the script

Let's not worry about this. If we find problems later we can always another test similar to https://github.com/tmeasday/create-graphql-server/tree/master/test/output-app-end-to-end to the intermediate states.

tobkle commented:
ok now?

tmeasday commented:
This is excellent, and I think we could merge it now (happy to if you'd prefer to get it in).

A couple of small things:

1. Is there any real purpose in having `update-type`? Using `add-type` a second time seems to behave the same. Would it be simpler to just keep that?

2. Should we make it so you type `remove-type Typename` rather than `remove-type path/to/input/file-that-may-no-longer-exist.graphql`?

tmeasday commented:
Also, would you like me to add you as a maintainer of the project?

tobkle commented:
removed update-type in favour of running add-type twice,
remove-type can be used also with just the type,

additionally: had to check also if user meant whether a type or a path name. So it will also react on the path name. If it is a path, it will find all its files.graphql recursively and will add/remove all. Added the test therefore as well. Maybe you have also a look on that.

tmeasday commented on README.md:
```diff
@@ -70,6 +72,34 @@ If the field references an array (again w/ or w/o nullability) of another type,
+
+create-graphql-server remove-type typename
+
+create-graphql-server remove-type path
```
Does this version remove all the types defined in `path`?

tobkle replied:
Yes it does

tmeasday commented:
This great, I'll merge -- and release if you are ready?
